lib/HLSL/LiveValueAnalysis.cpp

```diff
@@ -0,0 +1,1323 @@
+      PrettyStr << "Detected " << (int)detected << " live values but only " << (int)regs << " are used by metadata\n";
+      VSStr << "Detected " << (int)detected << " live values but only " << (int)regs << " are used by metadata\n";
+      for( auto I : DetectedInstr ) {
+        I->print( PrettyStr );
+        I->print( VSStr );
+        PrettyStr << "\n";
+        VSStr << "\n";
+      }
+      PrettyStr << "*********************\n";
+      VSStr << "*********************\n";
+    }
+    if( const llvm::DebugLoc &debugInfo = callSite->getDebugLoc() ) {
+      PrettyStr << "Total 32-bit registers: " << (int)regs << "\n\n";
+      VSStr << "Total 32-bit registers: " << (int)regs << "\n\n";
+    }
+
+    std::ostringstream TmpStr;
+    for( Instruction *I :
+      sortInstructions<Instruction *>( m_spillsPerTraceCall[callSite] ) ) {
+      TmpStr.str( "" );
+      {
+        if( isa<ExtractValueInst>( I ) ) {
+          I = dyn_cast<Instruction>(I->getOperand( 0 ));
+        }
+
+        if( I->isUsedByMetadata() ) {
+          unsigned int totalLineWidth = 0;
+
+          if( auto *L = LocalAsMetadata::getIfExists( I ) ) {
+            if( auto *MDV = MetadataAsValue::getIfExists( I->getContext(), L ) ) {
+              for( User *U : MDV->users() ) {
+                if( DbgValueInst *DVI = dyn_cast<DbgValueInst>(U) ) {
+
+                  TmpStr << translateValueToName( DVI, I );
+
+                  // Output USE locations
+                  std::vector<PHINode *> Phis;
+                  std::set<PHINode *> VisitedPhis;
+                  std::set<DILocation *> UseLocations;
+                  while( 1 ) {
+                    for( User *U : I->users() ) {
+                      if( auto Inst = dyn_cast<Instruction>(U) ) {
+                        if( DILocation *Loc = Inst->getDebugLoc() ) {
+                          // Manually search for duplication location;
+                          // DILocations can be distinct but their source
+                          // location is not
+                          bool bFound = false;
+                          DIScope *scope = Loc->getScope();
+                          for( auto L : UseLocations ) {
+                            if( L->getScope() == scope ) {
+                              if( L->getLine() == Loc->getLine() &&
+                                L->getColumn() == Loc->getColumn() ) {
+                                bFound = true;
+                                break;
+                              }
+                            }
+                          }
+
+                          if( !bFound ) {
+                            UseLocations.insert( Loc );
+
+                            // Traverse each control flow edge separately and
+                            // gather up seen defs along the way. By keeping
+                            // track of the from-block, we can select the
+                            // right incoming value at PHIs.
```

Comment:
So this is literally the same code as earlier. Make this a separate function.

Reply:
Refactored to a new function.